[Bryan Wee] iP

bryanwee023 opened the pull request:
# Individual Project Requirements

> "The aim of this task is to ensure that you are sufficiently familiar with the GFMD syntax." -CS2103 Website ([source](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html))

The following commits address the requirements needed from Week 2 onwards. 🍪 

* **Week 2**: Skeletal Duke, with basic Task functions (add, delete, mark & list tasks)
* **Week 3**: Refactored OO approach, Unit Tests, Added JavaDocs

Features to be added in Week 4:
1. Gradle
2. Checkstyle
3. Level 10
4. Varargs


More functions can be added in Duke::initCommandTable()
```java
private static Hashtable<String, Consumer<String>> initCommandTable() {

    Hashtable<String, Consumer<String>> ret = new Hashtable<>();
    ret.put(keyword, callback);  //keyword: cmd keyword, callback: function that accepts input line
    return ret;
}
```

Markdown Features:
- [X] Text formatting (eg. heading, bold, inline code etc.)
- [X] Lists (bullet, numbered & task)
- [X] Fenced code block (with syntax highlighting)
- [X] An emoji
- [X] Blockquote
- [X] A hyper link

bryanwee023 committed:
Add skeletal conversation

bryanwee023 committed:
Add saved list of tasks

bryanwee023 committed:
Add Task class

bryanwee023 committed:
Add error catching when marking Task

bryanwee023 committed:
Add different types of Task

bryanwee023 committed:
Add AddTask method

bryanwee023 committed:
Update automated tests

bryanwee023 committed:
Add DukeException

bryanwee023 committed:
Update automated tests

bryanwee023 committed:
Add deleting of tasks

bryanwee023 committed:
Clean up code

bryanwee023 committed:
Update automated tests

bryanwee023 committed:
Refactor control flow

bryanwee023 committed:
Fix keyword extraction bug

bryanwee023 committed:
Implement persistent data

bryanwee023 committed:
Represent time as LocalDate class

bryanwee023 committed:
Merge branch 'branch-Level-7'

bryanwee023 committed:
Fix index bug

bryanwee023 committed:
Abstract out storage

bryanwee023 committed:
Abstract out UI

bryanwee023 committed:
Abstract out Task list

bryanwee023 committed:
Add goodbye message

bryanwee023 committed:
Clean main flow

bryanwee023 committed:
Package files

bryanwee023 committed:
Add Equals method

bryanwee023 committed:
Add unit tests

bryanwee023 committed:
Allow flexible filepath

bryanwee023 committed:
Add META-INF to .gitignore

bryanwee023 committed:
Add JavaDocs

bryanwee023 committed:
Tweak to fit coding convention

bryanwee023 committed:
Add find command

bryanwee023 committed:
Merge branch 'branch-A-CodingStandard'

bryanwee023 committed:
Merge Level 9 branch

bryanwee023 committed:
Add gradle support

bryanwee023 committed:
Add checkstyle

bryanwee023 committed:
Fix code to checkstyle requirements

bryanwee023 committed:
Add JavaFX

bryanwee023 committed:
Remove text-based tests

pss-coder commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+
+    private static Hashtable<String, Consumer<String>> initCommandTable() {
+
+        Hashtable<String, Consumer<String>> ret = new Hashtable<>();
```
what does ret mean ? perhaps the name could mean something to what the object instance is trying to perform ?

pss-coder commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Scanner;
+import java.util.function.Consumer;
+
+public class Duke {
```
Could you add some header comments or comments in the class ? It would explain what the purpose of the class is.

pss-coder commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+
+        // Else, deserialize data
+        try {
+            FileInputStream f = new FileInputStream(DATA_FILE);
+            ObjectInputStream i = new ObjectInputStream(f);
```
could the names of the variable change to fit the naming convention ?

pss-coder commented on src/main/java/duke/TaskBank.java:
```diff
@@ -0,0 +1,102 @@
+        this.storage.writeToDisk(this.taskList);
+    }
+    
+    public ArrayList<Task> searchTasks(String query) {
```
Could comments be added to the method to follow the styling convention ?

pss-coder commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,75 @@
+
+public class Deadline extends Task {
+
+    private static final DateTimeFormatter DATE_SHORT_FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+    private static final DateTimeFormatter DATE_MED_FORMATTER = DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG);
```
I like the idea of reusable date format, great idea to have. Perhaps to reduce from code duplicating, would you consider storing all the reused date formatter into a utility class instead ?

pss-coder commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,81 @@
+     * @return Created Event
+     * @throws DukeException given string fails to meet format requirements
+     */
+    public static Event create(String formattedString) throws DukeException {
```
I understand what create() does, but could adding  Events at the name like createEvents() make it sound more accurate?

pss-coder reviewed (commented):
> Overall, LGTM. Might have a couple of naming convention nits to fix.

Hi Bryan, I was assigned to peer review your code. I really liked the idea in your Duke.java of using a HashTable to map userinput to a command. It also gave me 2030s and 2040s flashback😆 . 

It also inspired me to further re-assess my code. Thanks. 👍🏽

s7u4rt99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+
+    public static void main(String[] args) {
+
+        Ui.welcomeMessage();
+        Ui.printDivider();
+
+        Scanner sc=new Scanner(System.in);
+
+        Hashtable<String, Consumer<String>> commandTable = initCommandTable();
+
+        while (true) {
+            String input = sc.nextLine();
+            if (input.equals("bye"))
+                break;
+
+            String keyword = input.split(" ", 2)[0];
+            if (!commandTable.containsKey(keyword)) {
+                Ui.print("Oops, I'm not sure what you mean :o");
+                continue;
+            }
+
+            try {
+                commandTable.get(keyword).accept(input);
+            } catch (DukeException e) {
+                Ui.print(e.toString());
+            }
+
+            Ui.printDivider();
+        }
+
+        Ui.goodbyeMessage();
+    }
```
Really like the idea of storing the commands in a hashtable with O(1) access time which is very efficient. 
Perhaps it would be good to add comments to each part of the method so as to briefly describe what they are doing.

s7u4rt99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+        Ui.welcomeMessage();
+        Ui.printDivider();
+
+        Scanner sc=new Scanner(System.in);
```
Could u add some spaces here, so as to be consistent with the coding standard.
```suggestion
        Scanner sc = new Scanner(System.in);
```

s7u4rt99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.Collection;
+
+public class Ui {
+
+    private static final String DIVIDER = "------------------------------------------";
+    private static final String LOGO = " ____        _        \n"
+                                     + "|  _ \\ _   _| | _____ \n"
+                                     + "| | | | | | | |/ / _ \\\n"
+                                     + "| |_| | |_| |   <  __/\n"
+                                     + "|____/ \\__,_|_|\\_\\___|\n";
+
+    public static void welcomeMessage() {
+        System.out.println("\nHello buddy! I am");
+        System.out.println(LOGO);
+        System.out.println("What can I do for you?");
+    }
+
+    public static void goodbyeMessage() {
+        System.out.println("Bye. Hope to see you soon!");
+    }
+
+    public static void printDivider() {
+        System.out.println(DIVIDER);
+    }
+
+    public static void print(String text) {
+        System.out.println(text);
+    }
+
+    public static void printTasks(Collection<Task> tasks) {
+        int i = 0;
+        for (Task t: tasks) {
+            System.out.printf("%d. %s\n", i + 1, t);
+            i++;
+        }
+    }
```
I like the idea of having different methods to print different stuff. Perhaps by adding some header comments for each method and the class would make it easier to understand what each method does without the need to look at the implementation.

s7u4rt99 commented on src/main/java/duke/util/Utility.java:
```diff
@@ -0,0 +1,29 @@
+package duke.util;
+
```
Perhaps you can add some class header with a simple description of what this class is for.

s7u4rt99 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,75 @@
+
+    private static void checkFormat(String formattedString) throws DukeException {
+        int byIndex = formattedString.indexOf("/by ");
+        if (byIndex == -1)
+            byIndex = formattedString.length();
+
+        String keyword = formattedString.split(" ", 2)[0];
+
+        if (!keyword.startsWith("deadline"))
+            throw new DukeException("I can't seem to find the deadline keyword");
+        else if (formattedString.length() <= 9 || formattedString.substring(9, byIndex).isEmpty())
+            throw new DukeException("the description of deadline cannot be empty");
+        else if (byIndex == formattedString.length() || formattedString.length() < byIndex + 5)
+            throw new DukeException("the [/by] time of deadline cannot be empty");
```
I like the way you check for the keywords and handle all the different exceptions. Perhaps you can add Egyptian braces for each if/elseif block to be consistent with the coding standard.

s7u4rt99 reviewed (commented):
LGTM, just a couple of javadocs, comments and minor coding standard convention n its to fix.

bryanwee023 committed:
Add CSS styling

bryanwee023 committed:
Rework checkstyle

nhjryan reviewed (commented):
LGTM! All the code appears to adhere to the specified coding standards and have otherwise been suggested by others. :)

bryanwee023 committed:
Add assertions

bryanwee023 committed:
Enhance readability of code

Previously, certain portions of the code was difficult to decipher.
This might become a concern if the project becomes open to
collaboration.

Let's:
* Add comments into portions of code that are not self-explanatory
* Improve naming conventions to make them self-explanatory
* Shift date modifiers into the Utility class
* Decompose long lines into multiple short lines of code

These changes all aim to aid developers in understanding the code
easily. The changes keep code self-explanatory and minimising
repeated code, making code more reader-friendly.

bryanwee023 committed:
Merge pull request #2 from bryanwee023/A-CodeQuality

Code Quality Enhancements

bryanwee023 committed:
Merge pull request #1 from bryanwee023/A-Assertions

Add assertions

bryanwee023 committed:
Add test workflow on Github Actions

bryanwee023 committed:
Fix checkstyle to fit Gradle

bryanwee023 committed:
Fix coding style in test classes

bryanwee023 committed:
Merge pull request #3 from bryanwee023/A-CI

Add test workflow on Github Actions

bryanwee023 committed:
Add sort feature

When tracking many tasks, users might find it useful to be able to sort tasks.

Let's allow the sorting of tasks alphabetically and chronologically. The commands:
* ```list -name```: sorts and prints the tasks in alphabetical order
* ```list -time```: sorts and prints the tasks in chronological order
* ```list -done```: sorts and prints the tasks, printing undone tasks first

Looking ahead, we should be able to specify multiple sort orders.
e.g. sort -done -time sorts tasks by done status then by time.

bryanwee023 committed:
Merge pull request #4 from bryanwee023/C-Sort

Add sort feature

bryanwee023 committed:
delete SU_EDU checkstyle

bryanwee023 committed:
Edit status icon

bryanwee023 committed:
Restyle input bar

bryanwee023 committed:
Rework Duke personality

bryanwee023 committed:
Add title banner

bryanwee023 committed:
Fix clipping issue

bryanwee023 committed:
Add Ui.png

bryanwee023 committed:
Add user guide

bryanwee023 committed:
Set theme jekyll-theme-slate

bryanwee023 committed:
Catch bad date formatting

bryanwee023 committed:
Merge branch 'master' of https://github.com/bryanwee023/ip

bryanwee023 committed:
Delete su_edu.xml

bryanwee023 committed:
Merge branch 'master' of https://github.com/bryanwee023/ip

bryanwee023 committed:
Add Javadoc comments

bryanwee023 committed:
Add shadowjar plugin
